store/postgres/src/fork.rs

```diff
@@ -0,0 +1,347 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+};
+
+use graph::{
+    block_on,
+    components::store::SubgraphFork as SubgraphForkTrait,
+    prelude::{
+        info,
+        r::Value as RValue,
+        reqwest,
+        s::{Definition, Field, ObjectType, TypeDefinition},
+        serde_json, Attribute, DeploymentHash, Entity, Logger, Schema, Serialize, StoreError,
+        Value,
+    },
+    url::Url,
+};
+
+#[derive(Serialize, Debug, PartialEq)]
+struct Query {
+    query: String,
+    variables: Variables,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+struct Variables {
+    id: String,
+}
+
+/// SubgraphFork represents a simple subgraph forking mechanism
+/// which lazily fetches entities from a remote subgraph's store
+/// associated with a GraphQL endpoint at `fork_url`.
+///
+/// Since this mechanism is used for debug forks, entities are
+/// fetched only once per id in order to avoid fetching an entity
+/// that was deleted from the local store and thus causing inconsistencies.
+pub(crate) struct SubgraphFork {
+    client: reqwest::Client,
+    endpoint: Url,
+    schema: Arc<Schema>,
+    fetched_ids: Mutex<HashSet<String>>,
+    logger: Logger,
+}
+
+impl SubgraphForkTrait for SubgraphFork {
+    fn fetch(&self, entity_type: String, id: String) -> Result<Option<Entity>, StoreError> {
+        let mut fids = self.fetched_ids.lock().map_err(|e| {
+            StoreError::ForkFailure(format!(
+                "attempt to acquire lock on `fetched_ids` failed with {}",
+                e,
+            ))
+        })?;
+        if fids.contains(&id) {
+            info!(self.logger, "Already fetched entity! Abort!"; "entity_type" => entity_type, "id" => id);
+            return Ok(None);
+        }
+        fids.insert(id.clone());
```

Comment:
The lock should be dropped after this line. A good way to do this would be to encapsulate this mutex operation into a separate fn.

Reply:
What about just introducing a new scope?

Reply:
that also works